core/src/main/java/org/pentaho/di/connections/vfs/VFSConnectionManagerHelper.java

```diff
@@ -0,0 +1,372 @@
+   * @throws KettleException When a provider with the given key is not defined.
+   */
+  @NonNull
+  public <T extends VFSConnectionDetails> VFSConnectionProvider<T> getExistingProvider(
+    @NonNull ConnectionManager manager,
+    @NonNull T details )
+    throws KettleException {
+    return getExistingProvider( manager, details.getType() );
+  }
+  // endregion
+
+  // region Connection Root
+  @NonNull
+  public ConnectionFileName getConnectionRootFileName( @NonNull VFSConnectionDetails details ) {
+
+    String connectionName = details.getName();
+    if ( StringUtils.isEmpty( connectionName ) ) {
+      throw new IllegalArgumentException( "Unnamed connection" );
+    }
+
+    return new ConnectionFileName( encodeConnectionName( connectionName ) );
+  }
+
+  @NonNull
+  public <T extends VFSConnectionDetails> String getConnectionRootProviderUri(
+    @NonNull VFSConnectionFileNameTransformer<T> fileNameTransformer,
+    @NonNull T details )
+    throws KettleException {
+
+    // Example: "pvfs://my-connection"
+    ConnectionFileName rootPvfsFileName = getConnectionRootFileName( details );
+
+
+    // Example: "s3://root-bucket/root-path-subfolder"
+    return fileNameTransformer.toProviderUri( rootPvfsFileName, details );
+  }
+
+  @NonNull
+  private <T extends VFSConnectionDetails> FileObject getConnectionRootProviderFileObject(
+    @NonNull IKettleVFS kettleVFS,
+    @NonNull T details,
+    @NonNull VFSConnectionProvider<T> provider )
+    throws KettleException {
+
+    String rootPathProviderUri = getConnectionRootProviderUri( provider.getFileNameTransformer(), details );
+
+    return kettleVFS.getFileObject( rootPathProviderUri, new Variables(), provider.getOpts( details ) );
+  }
+  // endregion
+
+  /**
+   * Encodes the reserved characters in a given connection name.
+   *
+   * @param connectionName A connection name.
+   * @return A connection in canonical percent-encoding form.
+   * @see ConnectionFileNameParser#getReservedChars()
+   */
+  @NonNull
+  public String encodeConnectionName( @NonNull String connectionName ) {
+    return UriParser.encode( connectionName, fileNameParser.getReservedChars() );
+  }
+
+  // region VFS Resolved Root Path
+
+  /**
+   * Gets the resolved root path of a given connection.
+   * <p>
+   * This method returns the value of the root path after any variables substituted and subsequent normalization.
+   * <p>
+   * Root path normalization fixes separators, resolves "." and ".." segments, and removes any leading and trailing
+   * path separator.
+   * <p>
+   * Root path normalization does not affect the encoding of the original root path value, after variable
+   * substitution, which should be that of the specific connection/file provider.
+   *
+   * @param details The VFS connection details.
+   * @return The non-empty resolved root path, if any; {@code null}, if. none.
+   */
+  @Nullable
+  public String getResolvedRootPath( @NonNull VFSConnectionDetails details )
```

Comment:
The methods `test`, `getResolvedRootPath` and `useBuckets` were moved here from `VFSConnectionProvider`.